File: plugins/payment-contract-tester/reference/pytest/jwtmini.py
"""Minimal HS256 JWT encode/verify using only the standard library."""
import base64
import hashlib
import hmac
import json


def _b64u(b: bytes) -> str:
    return base64.urlsafe_b64encode(b).rstrip(b"=").decode()


def _b64u_dec(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def encode(claims: dict, secret: bytes) -> str:
    header = {"alg": "HS256", "typ": "JWT"}
    seg = _b64u(json.dumps(header, separators=(",", ":")).encode()) + "." + \
        _b64u(json.dumps(claims, separators=(",", ":")).encode())
    sig = hmac.new(secret, seg.encode(), hashlib.sha256).digest()
    return seg + "." + _b64u(sig)


def decode(token: str, secret: bytes) -> dict:
    """Verify the HS256 signature (constant-time) and return the claims."""
    seg, _, sig_b64 = token.rpartition(".")
    if not seg or not sig_b64:
        raise ValueError("malformed token")
    expected = hmac.new(secret, seg.encode(), hashlib.sha256).digest()
    if not hmac.compare_digest(expected, _b64u_dec(sig_b64)):
        raise ValueError("bad signature")
    _, _, payload_b64 = seg.partition(".")
    return json.loads(_b64u_dec(payload_b64))


def decode_unsafe(token: str) -> dict:
    """Parse claims WITHOUT verifying the signature. Used only by a seeded trap."""
    seg, _, _ = token.rpartition(".")
    _, _, payload_b64 = seg.partition(".")
    return json.loads(_b64u_dec(payload_b64))

File: plugins/payment-contract-tester/reference/xunit/JwtMini.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace ContractFixture;

/// <summary>Minimal HS256 JWT encode/verify using only the BCL (mirrors pytest jwtmini.py).</summary>
public static class JwtMini
{
    private static string B64u(byte[] b) =>
        Convert.ToBase64String(b).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    private static byte[] B64uDec(string s)
    {
        s = s.Replace('-', '+').Replace('_', '/');
        switch (s.Length % 4) { case 2: s += "=="; break; case 3: s += "="; break; }
        return Convert.FromBase64String(s);
    }

    public static string Encode(IDictionary<string, object?> claims, byte[] secret)
    {
        var header = JsonSerializer.SerializeToUtf8Bytes(
            new Dictionary<string, object?> { ["alg"] = "HS256", ["typ"] = "JWT" });
        var payload = JsonSerializer.SerializeToUtf8Bytes(claims);
        var seg = B64u(header) + "." + B64u(payload);
        var sig = HMACSHA256.HashData(secret, Encoding.UTF8.GetBytes(seg));
        return seg + "." + B64u(sig);
    }

    /// <summary>Verify the HS256 signature (constant-time) and return the claims.</summary>
    public static Dictionary<string, JsonElement> Decode(string token, byte[] secret)
    {
        var lastDot = token.LastIndexOf('.');
        if (lastDot <= 0) throw new FormatException("malformed token");
        var seg = token.Substring(0, lastDot);
        var sigB64 = token.Substring(lastDot + 1);
        if (seg.IndexOf('.') < 0 || sigB64.Length == 0) throw new FormatException("malformed token");
        var expected = HMACSHA256.HashData(secret, Encoding.UTF8.GetBytes(seg));
        if (!CryptographicOperations.FixedTimeEquals(expected, B64uDec(sigB64)))
            throw new FormatException("bad signature");
        var payloadB64 = seg.Substring(seg.IndexOf('.') + 1);
        return JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(B64uDec(payloadB64))!;
    }

    /// <summary>Parse claims WITHOUT verifying the signature. Used only by a seeded trap.</summary>
    public static Dictionary<string, JsonElement> DecodeUnsafe(string token)
    {
        var lastDot = token.LastIndexOf('.');
        var seg = token.Substring(0, lastDot);
        var payloadB64 = seg.Substring(seg.IndexOf('.') + 1);
        return JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(B64uDec(payloadB64))!;
    }
}
